llvm/lib/Transforms/Scalar/LICM.cpp

```diff
@@ -857,9 +859,18 @@ class ControlFlowHoister {
     }
 
     // Now finally clone BI.
-    ReplaceInstWithInst(
-        HoistTarget->getTerminator(),
-        BranchInst::Create(HoistTrueDest, HoistFalseDest, BI->getCondition()));
+    auto *NewBI =
+        BranchInst::Create(HoistTrueDest, HoistFalseDest, BI->getCondition(),
+                           HoistTarget->getTerminator()->getIterator());
+    HoistTarget->getTerminator()->eraseFromParent();
+    // Copy all the metadata. In particular:
+    // - debug info (critical to Sample-based profiling) should be the same
+    // as the original branch, not that of HoistTarget->getTerminator(),
+    // which is what ReplaceInstWithInst would use.
```

Comment:
Hm, looking at https://llvm.org/docs/HowToUpdateDebugInfo.html#when-to-preserve-an-instruction-location again, it specifically calls out LICM as a case where the debug location should be dropped, not preserved.

Reply:
Besides, in this case it wouldn't be misleading, for AutoFDO purposes either (or someone correct me if I'm missing something)

Reply:
Debug info usage of debuggers and SamplePGO is at odds here. Debuggers want this dropped, while SamplePGO does not. Our debug info policy is generally targeted towards debuggers.

Reply:
I think there's an existing flag that we could latch on to detect user intent, @snehasish? - i.e. if the intent is adding debug info for profiling, presumably we should bias towards that goal rather?

Reply:
Filed #152767 to track the addition of a new flag and to update the documentation. 

I don't think we should enable it by default for O2/O3 though. I received pushback even when suggesting we add it under `-fdebug-info-for-profiling`. It should be purely opt-in especially for the cases where the location may be non-representable.

To keep this PR moving along can we leave the behaviour for debug information unchanged (add a TODO with the issue number) and only address the profile metadata being dropped?

Reply:
You'll still want to explicitly set the location to dropped here to not run afoul of verification (https://discourse.llvm.org/t/rfc-require-real-or-annotated-source-locations-on-all-instructions/86816).